ykrt/src/location.rs

```diff
@@ -97,88 +29,407 @@ impl State {
+        debug_assert!(!ls.is_counting());
+        if self
+            .state
+            .compare_exchange(
+                ls.with_unparked().x,
+                ls.with_unparked().with_unlock().x,
+                Ordering::Release,
+                Ordering::Relaxed,
+            )
+            .is_ok()
+        {
+            return;
+        }
+
+        // At this point we know another thread has parked itself.
+        let key = unsafe { ls.hot_location() } as *const _ as usize;
+        debug_assert_ne!(key, 0);
+        let callback = |result: UnparkResult| {
+            if result.unparked_threads != 0 && result.be_fair {
+                if !result.have_more_threads {
+                    debug_assert!(ls.is_locked());
+                    self.state.store(ls.with_unparked().x, Ordering::Relaxed);
+                }
+                return TOKEN_HANDOFF;
+            }
+
+            if result.have_more_threads {
+                self.state
+                    .store(ls.with_unlock().with_parked().x, Ordering::Release);
+            } else {
+                self.state
+                    .store(ls.with_unparked().with_unlock().x, Ordering::Release);
+            }
+            TOKEN_NORMAL
+        };
+        unsafe {
+            unpark_one(key, callback);
+        }
     }
 
-    pub(crate) fn store(&self, state: State, order: Ordering) {
-        self.state.store(state.0, order);
+    /// Try obtaining the lock, returning the new `State` if successful.
+    pub(crate) fn try_lock(&self) -> Result<State<I>, ()> {
+        let mut ls = self.load(Ordering::Relaxed);
+        loop {
+            if ls.is_locked() {
+                return Err(());
+            }
+            let new_ls = ls.with_lock();
+            match self.compare_exchange_weak(ls, new_ls, Ordering::Acquire, Ordering::Relaxed) {
+                Ok(_) => return Ok(new_ls),
+                Err(x) => ls = x,
+            }
+        }
     }
 }
 
 impl<I> Drop for Location<I> {
     fn drop(&mut self) {
-        let lp = *self.state.get_mut();
-        if lp & PHASE_TAG == PHASE_TRACING {
-            unsafe { Arc::from_raw((lp & !PHASE_TAG) as *mut u8) };
+        let ls = self.load(Ordering::Relaxed);
+        if !ls.is_counting() {
+            debug_assert!(!ls.is_locked());
+            let hr = unsafe { ls.hot_location() };
+            unsafe {
+                Box::from_raw(hr);
+            }
+        }
+    }
+}
+
+#[cfg(target_pointer_width = "64")]
+pub(crate) const STATE_TAG: usize = 0b111; // All of the other tag data must fit in this.
+#[cfg(target_pointer_width = "64")]
+pub(crate) const STATE_NUM_BITS: usize = 3;
+
+pub(crate) const STATE_IS_LOCKED: usize = 0b001;
+pub(crate) const STATE_IS_PARKED: usize = 0b010;
+pub(crate) const STATE_IS_COUNTING: usize = 0b100;
+
+pub(crate) const TOKEN_NORMAL: UnparkToken = UnparkToken(0);
+pub(crate) const TOKEN_HANDOFF: UnparkToken = UnparkToken(1);
+
+#[derive(PartialEq, Eq, Debug)]
```

Comment:
Please add docstrings. What is `x`?

Reply:
This would be called `0` if it wasn't for the `PhantomType`. I don't think naming this or adding a doc string is very useful.

Reply:
Why don't you hide it behind a `Deref`? Both @ptersilie and I were confused by `x`.

Reply:
I realised that the `x` Lukas is questioning is a different `x`!

Reply:
In the long run, I'm not very happy with the relationship between `Location` and `State` and I want to try tidying it up. I don't want to get too bogged down in this now though.